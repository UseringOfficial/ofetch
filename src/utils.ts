import type {
  FetchContext,
  FetchHookNext,
  FetchHooks,
  FetchOptions,
  ResponseType,
} from "./types";

const payloadMethods = new Set(
  Object.freeze(["PATCH", "POST", "PUT", "DELETE"])
);
export function isPayloadMethod(method = "GET") {
  return payloadMethods.has(method.toUpperCase());
}

export function isJSONSerializable(value: any) {
  if (value === undefined) {
    return false;
  }
  const t = typeof value;
  if (t === "string" || t === "number" || t === "boolean" || t === null) {
    return true;
  }
  if (t !== "object") {
    return false; // bigint, function, symbol, undefined
  }
  if (Array.isArray(value)) {
    return true;
  }
  if (value.buffer) {
    return false;
  }
  return (
    (value.constructor && value.constructor.name === "Object") ||
    typeof value.toJSON === "function"
  );
}

const textTypes = new Set([
  "image/svg",
  "application/xml",
  "application/xhtml",
  "application/html",
]);

const JSON_RE = /^application\/(?:[\w!#$%&*.^`~-]*\+)?json(;.+)?$/i;

// This provides reasonable defaults for the correct parser based on Content-Type header.
export function detectResponseType(_contentType = ""): ResponseType {
  if (!_contentType) {
    return "json";
  }

  // Value might look like: `application/json; charset=utf-8`
  const contentType = _contentType.split(";").shift() || "";

  if (JSON_RE.test(contentType)) {
    return "json";
  }

  // TODO
  // if (contentType === 'application/octet-stream') {
  //   return 'stream'
  // }

  if (textTypes.has(contentType) || contentType.startsWith("text/")) {
    return "text";
  }

  return "blob";
}

export const FETCH_HOOKS: ReadonlyArray<keyof FetchHooks> = [
  "onRequest",
  "onRequestError",
  "onResponse",
  "onResponseError",
  "hasBody",
];

// Merging of fetch option objects.
export function mergeFetchOptions(
  input: FetchOptions | undefined,
  defaults: FetchOptions | undefined,
  Headers = globalThis.Headers
): FetchOptions {
  const merged: FetchOptions = {
    ...defaults,
    ...input,
  };

  // Merge params and query
  if (defaults?.params && input?.params) {
    merged.params = {
      ...defaults?.params,
      ...input?.params,
    };
  }
  if (defaults?.query && input?.query) {
    merged.query = {
      ...defaults?.query,
      ...input?.query,
    };
  }

  // Merge headers
  if (defaults?.headers && input?.headers) {
    merged.headers = new Headers(defaults?.headers || {});
    for (const [key, value] of new Headers(input?.headers || {})) {
      merged.headers.set(key, value);
    }
  }

  for (const name of FETCH_HOOKS) {
    merged[name] = (async (context, next) => {
      return input?.[name]
        ? await input[name](
            context as any,
            async () => (await defaults?.[name]?.(context as any, next)) as any
          )
        : await defaults?.[name]?.(context as any, next);
    }) as (context: FetchContext, next: FetchHookNext) => any;
  }

  return merged;
}

export async function noop(): Promise<void> {
  // noop
}
