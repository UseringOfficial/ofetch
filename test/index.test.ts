import {
  createApp,
  createError,
  eventHandler,
  readBody,
  readRawBody,
  toNodeListener,
} from "h3";
import { listen, type Listener } from "listhen";
import { Blob, FormData, Headers } from "node-fetch-native";
import { Readable } from "node:stream";
import * as qs from "qs";
import { nodeMajorVersion } from "std-env";
import { getQuery, joinURL, type QueryObject } from "ufo";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { R } from "vitest/dist/reporters-yx5ZTtEV";
import {
  $fetch,
  createFetch,
  type FetchContext,
  type FetchHookNext,
  type FetchHooks,
  type FetchResponse,
} from "../src/node";
import { noop } from "../src/utils";

describe("ofetch", () => {
  let listener: Listener;
  const getURL = (url: string) => joinURL(listener.url, url);

  beforeAll(async () => {
    const app = createApp()
      .use(
        "/ok",
        eventHandler(() => "ok")
      )
      .use(
        "/params",
        eventHandler((event) => getQuery(event.node.req.url || ""))
      )
      .use(
        "/url",
        eventHandler((event) => event.node.req.url)
      )
      .use(
        "/echo",
        eventHandler(async (event) => ({
          path: event.path,
          body:
            event.node.req.method === "POST"
              ? await readRawBody(event)
              : undefined,
          headers: event.node.req.headers,
        }))
      )
      .use(
        "/post",
        eventHandler(async (event) => ({
          body: await readBody(event),
          headers: event.node.req.headers,
        }))
      )
      .use(
        "/binary",
        eventHandler((event) => {
          event.node.res.setHeader("Content-Type", "application/octet-stream");
          return new Blob(["binary"]);
        })
      )
      .use(
        "/403",
        eventHandler(() =>
          createError({ status: 403, statusMessage: "Forbidden" })
        )
      )
      .use(
        "/408",
        eventHandler(() => createError({ status: 408 }))
      )
      .use(
        "/204",
        eventHandler(() => null) // eslint-disable-line unicorn/no-null
      )
      .use(
        "/timeout",
        eventHandler(async () => {
          await new Promise((resolve) => {
            setTimeout(() => {
              resolve(createError({ status: 408 }));
            }, 1000 * 5);
          });
        })
      );

    listener = await listen(toNodeListener(app), {
      hostname: "localhost",
    });
  });

  afterAll(() => {
    listener.close().catch(console.error);
  });

  it("ok", async () => {
    expect(await $fetch(getURL("ok"))).to.equal("ok");
  });

  it("custom parseResponse", async () => {
    let called = 0;
    const parser = async (
      context: FetchContext & { response: FetchResponse<R> }
    ) => {
      called++;
      return "C" + (await context.response.text());
    };
    expect(await $fetch(getURL("ok"), { parseResponse: parser })).to.equal(
      "Cok"
    );
    expect(called).to.equal(1);
  });

  it("allows specifying FetchResponse method", async () => {
    expect(
      await $fetch(getURL("params?test=true"), { responseType: "json" })
    ).to.deep.equal({ test: "true" });
    expect(
      await $fetch(getURL("params?test=true"), { responseType: "blob" })
    ).to.be.instanceOf(Blob);
    expect(
      await $fetch(getURL("params?test=true"), { responseType: "text" })
    ).to.equal('{"test":"true"}');
    expect(
      await $fetch(getURL("params?test=true"), { responseType: "arrayBuffer" })
    ).to.be.instanceOf(ArrayBuffer);
  });

  it("returns a blob for binary content-type", async () => {
    expect(await $fetch(getURL("binary"))).to.be.instanceOf(Blob);
  });

  it("baseURL", async () => {
    expect(await $fetch("/x?foo=123", { baseURL: getURL("url") })).to.equal(
      "/x?foo=123"
    );
  });

  it("stringifies posts body automatically", async () => {
    const { body } = await $fetch(getURL("post"), {
      method: "POST",
      body: { num: 42 },
    });
    expect(body).to.deep.eq({ num: 42 });

    const body2 = (
      await $fetch(getURL("post"), {
        method: "POST",
        body: [{ num: 42 }, { num: 43 }],
      })
    ).body;
    expect(body2).to.deep.eq([{ num: 42 }, { num: 43 }]);

    const headerFetches = [
      [["X-header", "1"]],
      { "x-header": "1" },
      new Headers({ "x-header": "1" }),
    ];

    for (const sentHeaders of headerFetches) {
      const { headers } = await $fetch(getURL("post"), {
        method: "POST",
        body: { num: 42 },
        headers: sentHeaders as HeadersInit,
      });
      expect(headers).to.include({ "x-header": "1" });
      expect(headers).to.include({ accept: "application/json" });
    }
  });

  it("does not stringify body when content type != application/json", async () => {
    const message = '"Hallo von Pascal"';
    const { body } = await $fetch(getURL("echo"), {
      method: "POST",
      body: message,
      headers: { "Content-Type": "text/plain" },
    });
    expect(body).to.deep.eq(message);
  });

  it("Handle Buffer body", async () => {
    const message = "Hallo von Pascal";
    const { body } = await $fetch(getURL("echo"), {
      method: "POST",
      body: Buffer.from("Hallo von Pascal"),
      headers: { "Content-Type": "text/plain" },
    });
    expect(body).to.deep.eq(message);
  });

  it.skipIf(Number(nodeMajorVersion) < 18)(
    "Handle ReadableStream body",
    async () => {
      const message = "Hallo von Pascal";
      const { body } = await $fetch(getURL("echo"), {
        method: "POST",
        headers: {
          "content-length": "16",
        },
        body: new ReadableStream({
          start(controller) {
            controller.enqueue(new TextEncoder().encode(message));
            controller.close();
          },
        }),
      });
      expect(body).to.deep.eq(message);
    }
  );

  it.skipIf(Number(nodeMajorVersion) < 18)("Handle Readable body", async () => {
    const message = "Hallo von Pascal";
    const { body } = await $fetch(getURL("echo"), {
      method: "POST",
      headers: {
        "content-length": "16",
      },
      body: new Readable({
        read() {
          this.push(message);
          this.push(null); // eslint-disable-line unicorn/no-null
        },
      }),
    });
    expect(body).to.deep.eq(message);
  });

  it("Bypass FormData body", async () => {
    const data = new FormData();
    data.append("foo", "bar");
    const { body } = await $fetch(getURL("post"), {
      method: "POST",
      body: data,
    });
    expect(body).to.include('form-data; name="foo"');
  });

  it("Bypass URLSearchParams body", async () => {
    const data = new URLSearchParams({ foo: "bar" });
    const { body } = await $fetch(getURL("post"), {
      method: "POST",
      body: data,
    });
    expect(body).toMatchObject({ foo: "bar" });
  });

  it("404", async () => {
    const error = await $fetch(getURL("404")).catch((error_) => error_);
    expect(error.toString()).to.contain("Cannot find any path matching /404.");
    expect(error.data).to.deep.eq({
      stack: [],
      statusCode: 404,
      statusMessage: "Cannot find any path matching /404.",
    });
    expect(error.response?._data).to.deep.eq(error.data);
    expect(error.request).to.equal(getURL("404"));
  });

  it("403 with ignoreResponseError", async () => {
    const res = await $fetch(getURL("403"), { ignoreResponseError: true });
    expect(res?.statusCode).to.eq(403);
    expect(res?.statusMessage).to.eq("Forbidden");
  });

  it("204 no content", async () => {
    const res = await $fetch(getURL("204"));
    expect(res).toBeUndefined();
  });

  it("HEAD no content", async () => {
    const res = await $fetch(getURL("/ok"), { method: "HEAD" });
    expect(res).toBeUndefined();
  });

  it("baseURL with retry", async () => {
    const error = await $fetch("", { baseURL: getURL("404"), retry: 3 }).catch(
      (error_) => error_
    );
    expect(error.request).to.equal(getURL("404"));
  });

  it("retry with number delay", async () => {
    const slow = $fetch<string>(getURL("408"), {
      retry: 2,
      retryDelay: 100,
    }).catch(() => "slow");
    const fast = $fetch<string>(getURL("408"), {
      retry: 2,
      retryDelay: 1,
    }).catch(() => "fast");

    const race = await Promise.race([slow, fast]);
    expect(race).to.equal("fast");
  });

  it("retry with callback delay", async () => {
    const slow = $fetch<string>(getURL("408"), {
      retry: 2,
      retryDelay: () => 100,
    }).catch(() => "slow");
    const fast = $fetch<string>(getURL("408"), {
      retry: 2,
      retryDelay: () => 1,
    }).catch(() => "fast");

    const race = await Promise.race([slow, fast]);
    expect(race).to.equal("fast");
  });

  it("abort with retry", () => {
    const controller = new AbortController();
    async function abortHandle() {
      controller.abort();
      const response = await $fetch("", {
        baseURL: getURL("ok"),
        retry: 3,
        signal: controller.signal,
      });
      console.log("response", response);
    }
    expect(abortHandle()).rejects.toThrow(/aborted/);
  });

  it("passing request obj should return request obj in error", async () => {
    const error = await $fetch(getURL("/403"), { method: "post" }).catch(
      (error) => error
    );
    expect(error.toString()).toBe(
      'FetchError: [POST] "http://localhost:3000/403": 403 Forbidden'
    );
    expect(error.request).to.equal(getURL("403"));
    expect(error.options.method).to.equal("POST");
    expect(error.response?._data).to.deep.eq(error.data);
  });

  it("aborting on timeout", async () => {
    const noTimeout = $fetch(getURL("timeout")).catch(() => "no timeout");
    const timeout = $fetch(getURL("timeout"), {
      timeout: 100,
      retry: 0,
    }).catch(() => "timeout");
    const race = await Promise.race([noTimeout, timeout]);
    expect(race).to.equal("timeout");
  });

  it("aborting on timeout reason", async () => {
    await $fetch(getURL("timeout"), {
      timeout: 100,
      retry: 0,
    }).catch((error) => {
      expect(error.cause.message).to.include(
        "The operation was aborted due to timeout"
      );
      expect(error.cause.name).to.equal("TimeoutError");
      expect(error.cause.code).to.equal(DOMException.TIMEOUT_ERR);
    });
  });

  it("deep merges defaultOptions", async () => {
    const _customFetch = $fetch.create({
      query: {
        a: 0,
      },
      params: {
        b: 2,
      },
      headers: {
        "x-header-a": "0",
        "x-header-b": "2",
      },
    });
    const { headers, path } = await _customFetch(getURL("echo"), {
      query: {
        a: 1,
      },
      params: {
        c: 3,
      },
      headers: {
        "Content-Type": "text/plain",
        "x-header-a": "1",
        "x-header-c": "3",
      },
    });

    expect(headers).to.include({
      "x-header-a": "1",
      "x-header-b": "2",
      "x-header-c": "3",
    });

    expect(path).to.eq("?b=2&c=3&a=1");
  });

  it("merge hooks", async () => {
    const baseHooks: FetchHooks = {
      onRequest: vi.fn(),
      onResponse: vi.fn(),
      onResponseError: vi.fn(),
    };

    const base = createFetch({
      defaults: baseHooks,
    });

    const customize: FetchHooks = {
      onRequest: vi.fn(async (_context, next) => {
        await next();
      }),
      onResponse: vi.fn(async (_context, next) => {
        await next();
      }),
      onResponseError: vi.fn(async (_context, next) => {
        await next();
      }),
    };

    const customFetch = base.create(customize);

    await customFetch(getURL("403")).catch(noop);

    for (const name of Object.keys(customize) as Array<keyof FetchHooks>) {
      expect(customize[name]).toBeCalled();
      expect(baseHooks[name]).toBeCalled();
    }
  });

  it("check error set by onResponse", async () => {
    const message = "custom error";

    const error = await $fetch(getURL("ok"), {
      onResponse(
        context: FetchContext & { response: FetchResponse<R> },
        next: FetchHookNext
      ): Promise<void> | void {
        next();
        context.error = new Error(message);
      },
    }).catch((error) => error);

    expect(error.toString()).toBe(
      'FetchError: [GET] "http://localhost:3000/ok": 200 OK custom error'
    );
  });

  it("custom stringify query", async () => {
    const value = await $fetch(getURL("params"), {
      query: {
        foo: {
          bar: {
            baz: "value",
          },
        },
      },
      stringifyQuery(query: QueryObject): string {
        return qs.stringify(query);
      },
    });
    expect(value).toMatchSnapshot();
  });
});
