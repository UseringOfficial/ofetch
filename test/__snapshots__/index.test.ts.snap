// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ofetch > custom stringify query 1`] = `
{
  "foo[bar][baz]": "value",
}
`;
